code/obj/machinery/plantpot.dm

```diff
@@ -706,8 +749,20 @@
 
 	// Procs specific to the plantpot start here.
 
+	proc/update_water()
+		src.water_sprite = image('icons/obj/hydroponics/hydroponics.dmi',"wat-[src.total_volume]")
+		src.water_display = image('icons/obj/hydroponics/hydroponics.dmi',"ind-wat-[src.water_level]")
+
+		if (src.reagents.total_volume)
+			var/datum/color/average = src.reagents.get_average_color()
+			src.water_sprite.color = average.to_rgba()
+
+		UpdateOverlays(src.water_display, "water")
+		UpdateOverlays(src.water_sprite, "water_fluid")
```

Comment:
Consolidate into `update_icon()` proc.

Reply:
i personally removed it from update icon because the water state wasnt updating on reagent change. so it updates on reagent change separately, but other things trigger that update it too, so it also updates on plant icon update. Because of this duality it needs to be a separate proc. And even so. not cramming everything into a single proc is better code practice anyway. especially for opensource.

Reply:
I would recommend:

- A new `calculate_reagent_contents()` proc which does all of the `current_total_volume` and `current_water_level` stuff (lines 146-168) `on_reagent_change()` currently does, also add a `do_update_icon = 1` right after, in the same scope as `src.water_level = current_water_level`
- On `on_reagent_change()`, call `calculate_reagent_contents()` and then `update_icon()` 
- Remove `current_total_volume` and `current_water_level` stuff (lines 200-223) from `process()` and replace with a `calculate_reagents_contents()` call
- Remove `update_water()` and move its color and overlay stuff (lines 756-758, 761) to `update_icon()`, at the start, before `if (!src.current)` (line 767)

The first three points help avoid duplicate code by turning it into a proc. For the fourth point, because `update_water()` is only called once and shares similar purposes (and bits of code) with `update_icon()`, it'd be good to consolidate the former into the latter. If you feel like that'd make `update_icon()` too unwieldy or bloated, as an alternative to points two and four, you could also:

- On `on_reagent_change()`, call `calculate_reagent_contents()` and then `update_water()` 
- Move the water overlay stuff of `update_icon()` (lines 764-766) to `update_water()`, and then just call `update_water()` as your first thing in `update_icon()`

Reply:
ooh i really like that. I'll do another pass over the code tomorrow :heart:

Reply:
Has this been resolved?

Reply:
yep :)